File: apps/reframe/backend/app/services/groq_monitor.py
"""
Groq Usage Monitoring for Reframe
Tracks token usage and costs, stores in Redis
"""
import os
from datetime import datetime
from typing import Optional
from .redis_service import get_redis_service


# Groq pricing for llama-3.1-8b-instant (Reframe model)
GROQ_INPUT_COST_PER_MILLION = 0.05  # $0.05 per 1M input tokens
GROQ_OUTPUT_COST_PER_MILLION = 0.08  # $0.08 per 1M output tokens

# Alert thresholds (monthly only)
MONTHLY_COST_THRESHOLD = float(os.getenv("REFRAME_GROQ_MONTHLY_COST_THRESHOLD", os.getenv("GROQ_MONTHLY_COST_THRESHOLD", "50.0")))  # $50/month


def calculate_groq_cost(input_tokens: int, output_tokens: int) -> float:
    """Calculate cost in USD for Groq API usage"""
    input_cost = (input_tokens / 1_000_000) * GROQ_INPUT_COST_PER_MILLION
    output_cost = (output_tokens / 1_000_000) * GROQ_OUTPUT_COST_PER_MILLION
    return input_cost + output_cost


async def track_groq_usage(
    input_tokens: int,
    output_tokens: int,
    request_type: str = "reframe"
) -> None:
    """Track Groq API usage in Redis"""
    try:
        redis = get_redis_service()
        now = datetime.utcnow()
        date_key = now.strftime("%Y-%m-%d")  # YYYY-MM-DD
        month_key = f"{now.year}-{now.month:02d}"  # YYYY-MM
        
        total_tokens = input_tokens + output_tokens
        cost = calculate_groq_cost(input_tokens, output_tokens)
        
        # Track daily usage
        daily_key = f"groq:usage:daily:{date_key}"
        await redis.incrby(f"{daily_key}:requests", 1)
        await redis.incrby(f"{daily_key}:input_tokens", input_tokens)
        await redis.incrby(f"{daily_key}:output_tokens", output_tokens)
        await redis.incrbyfloat(f"{daily_key}:cost", cost)
        
        # Track monthly usage
        monthly_key = f"groq:usage:monthly:{month_key}"
        await redis.incrby(f"{monthly_key}:requests", 1)
        await redis.incrby(f"{monthly_key}:input_tokens", input_tokens)
        await redis.incrby(f"{monthly_key}:output_tokens", output_tokens)
        await redis.incrbyfloat(f"{monthly_key}:cost", cost)
        
        # Set expiration (30 days for daily, 90 days for monthly)
        await redis.expire(daily_key, 30 * 24 * 60 * 60)
        await redis.expire(monthly_key, 90 * 24 * 60 * 60)
        
        # Store individual request for detailed tracking
        import random
        import time
        request_key = f"groq:request:{int(time.time() * 1000)}:{random.randint(100000, 999999)}"
        await redis.hset(request_key, "input_tokens", str(input_tokens))
        await redis.hset(request_key, "output_tokens", str(output_tokens))
        await redis.hset(request_key, "total_tokens", str(total_tokens))
        await redis.hset(request_key, "cost", str(cost))
        await redis.hset(request_key, "request_type", request_type)
        await redis.hset(request_key, "timestamp", now.isoformat())
        await redis.expire(request_key, 30 * 24 * 60 * 60)  # 30 days
        
        # Check for alerts
        await _check_and_alert(redis, date_key, month_key)
    except Exception as error:
        print(f"Failed to track Groq usage: {error}")
        # Don't throw - tracking failure shouldn't break the request


async def _check_and_alert(redis, date_key: str, month_key: str) -> None:
    """Check usage and send alerts if thresholds exceeded"""
    try:
        daily_key = f"groq:usage:daily:{date_key}"
        monthly_key = f"groq:usage:monthly:{month_key}"
        
        monthly_cost_str = await redis.get(f"{monthly_key}:cost")
        monthly_cost = float(monthly_cost_str) if monthly_cost_str else 0.0
        
        # Check monthly threshold
        if monthly_cost > MONTHLY_COST_THRESHOLD:
            alert_key = f"groq:alert:monthly:{month_key}"
            already_alerted = await redis.get(alert_key)
            
            if not already_alerted:
                print(f"[GROQ ALERT] Monthly cost (${monthly_cost:.2f}) exceeds threshold (${MONTHLY_COST_THRESHOLD})")
                await redis.set(alert_key, "1")
                await redis.expire(alert_key, 30 * 24 * 60 * 60)  # Alert once per month
    except Exception as error:
        print(f"Failed to check alerts: {error}")

File: lib/reframe/groq-monitor.ts
/**
 * Groq Usage Monitoring for Reframe
 * Tracks token usage and costs, stores in Redis
 */

import { getRedisClient } from "./redis";

// Groq pricing for llama-3.1-8b-instant (Reframe model)
const GROQ_INPUT_COST_PER_MILLION = 0.05; // $0.05 per 1M input tokens
const GROQ_OUTPUT_COST_PER_MILLION = 0.08; // $0.08 per 1M output tokens

// Alert thresholds
const DAILY_COST_THRESHOLD = parseFloat(process.env.REFRAME_GROQ_DAILY_COST_THRESHOLD || process.env.GROQ_DAILY_COST_THRESHOLD || "10.0"); // $10/day
const MONTHLY_COST_THRESHOLD = parseFloat(process.env.REFRAME_GROQ_MONTHLY_COST_THRESHOLD || process.env.GROQ_MONTHLY_COST_THRESHOLD || "50.0"); // $50/month

/**
 * Calculate cost in USD for Groq API usage
 */
export function calculateGroqCost(inputTokens: number, outputTokens: number): number {
  const inputCost = (inputTokens / 1_000_000) * GROQ_INPUT_COST_PER_MILLION;
  const outputCost = (outputTokens / 1_000_000) * GROQ_OUTPUT_COST_PER_MILLION;
  return inputCost + outputCost;
}

/**
 * Track Groq API usage in Redis
 */
export async function trackGroqUsage(
  inputTokens: number,
  outputTokens: number,
  requestType: string = "reframe"
): Promise<void> {
  try {
    const redis = getRedisClient();
    const now = new Date();
    const dateKey = now.toISOString().split('T')[0]; // YYYY-MM-DD
    const monthKey = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}`; // YYYY-MM
    
    const totalTokens = inputTokens + outputTokens;
    const cost = calculateGroqCost(inputTokens, outputTokens);
    
    // Track daily usage
    const dailyKey = `groq:usage:daily:${dateKey}`;
    await redis.incrby(`${dailyKey}:requests`, 1);
    await redis.incrby(`${dailyKey}:input_tokens`, inputTokens);
    await redis.incrby(`${dailyKey}:output_tokens`, outputTokens);
    await redis.incrbyfloat(`${dailyKey}:cost`, cost);
    
    // Track monthly usage
    const monthlyKey = `groq:usage:monthly:${monthKey}`;
    await redis.incrby(`${monthlyKey}:requests`, 1);
    await redis.incrby(`${monthlyKey}:input_tokens`, inputTokens);
    await redis.incrby(`${monthlyKey}:output_tokens`, outputTokens);
    await redis.incrbyfloat(`${monthlyKey}:cost`, cost);
    
    // Set expiration (30 days for daily, 90 days for monthly)
    await redis.expire(dailyKey, 30 * 24 * 60 * 60);
    await redis.expire(monthlyKey, 90 * 24 * 60 * 60);
    
    // Store individual request for detailed tracking
    const requestKey = `groq:request:${Date.now()}:${Math.random().toString(36).substr(2, 9)}`;
    await redis.hset(requestKey, {
      input_tokens: String(inputTokens),
      output_tokens: String(outputTokens),
      total_tokens: String(totalTokens),
      cost: String(cost),
      request_type: requestType,
      timestamp: now.toISOString()
    });
    await redis.expire(requestKey, 30 * 24 * 60 * 60); // 30 days
    
    // Check for alerts
    await checkAndAlert(redis, dateKey, monthKey);
  } catch (error) {
    console.error("Failed to track Groq usage:", error);
    // Don't throw - tracking failure shouldn't break the request
  }
}

/**
 * Get daily usage statistics
 */
export async function getDailyUsage(date?: string): Promise<{
  date: string;
  requests: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  cost: number;
}> {
  try {
    const redis = getRedisClient();
    const dateKey = date || new Date().toISOString().split('T')[0];
    const dailyKey = `groq:usage:daily:${dateKey}`;
    
    const requests = parseInt((await redis.get(`${dailyKey}:requests`)) || "0", 10);
    const inputTokens = parseInt((await redis.get(`${dailyKey}:input_tokens`)) || "0", 10);
    const outputTokens = parseInt((await redis.get(`${dailyKey}:output_tokens`)) || "0", 10);
    const cost = parseFloat((await redis.get(`${dailyKey}:cost`)) || "0");
    
    return {
      date: dateKey,
      requests,
      inputTokens,
      outputTokens,
      totalTokens: inputTokens + outputTokens,
      cost: Math.round(cost * 1000000) / 1000000 // Round to 6 decimal places
    };
  } catch (error) {
    console.error("Failed to get daily usage:", error);
    return {
      date: date || new Date().toISOString().split('T')[0],
      requests: 0,
      inputTokens: 0,
      outputTokens: 0,
      totalTokens: 0,
      cost: 0
    };
  }
}

/**
 * Get monthly usage statistics
 */
export async function getMonthlyUsage(year?: number, month?: number): Promise<{
  year: number;
  month: number;
  requests: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  cost: number;
}> {
  try {
    const redis = getRedisClient();
    const now = new Date();
    const yearNum = year || now.getFullYear();
    const monthNum = month || now.getMonth() + 1;
    const monthKey = `${yearNum}-${String(monthNum).padStart(2, '0')}`;
    const monthlyKey = `groq:usage:monthly:${monthKey}`;
    
    const requests = parseInt((await redis.get(`${monthlyKey}:requests`)) || "0", 10);
    const inputTokens = parseInt((await redis.get(`${monthlyKey}:input_tokens`)) || "0", 10);
    const outputTokens = parseInt((await redis.get(`${monthlyKey}:output_tokens`)) || "0", 10);
    const cost = parseFloat((await redis.get(`${monthlyKey}:cost`)) || "0");
    
    return {
      year: yearNum,
      month: monthNum,
      requests,
      inputTokens,
      outputTokens,
      totalTokens: inputTokens + outputTokens,
      cost: Math.round(cost * 1000000) / 1000000
    };
  } catch (error) {
    console.error("Failed to get monthly usage:", error);
    const now = new Date();
    return {
      year: year || now.getFullYear(),
      month: month || now.getMonth() + 1,
      requests: 0,
      inputTokens: 0,
      outputTokens: 0,
      totalTokens: 0,
      cost: 0
    };
  }
}

/**
 * Check usage and send alerts if thresholds exceeded
 */
async function checkAndAlert(redis: any, dateKey: string, monthKey: string): Promise<void> {
  try {
    const dailyKey = `groq:usage:daily:${dateKey}`;
    const monthlyKey = `groq:usage:monthly:${monthKey}`;
    
    const dailyCost = parseFloat((await redis.get(`${dailyKey}:cost`)) || "0");
    const monthlyCost = parseFloat((await redis.get(`${monthlyKey}:cost`)) || "0");
    
    // Check daily threshold
    if (dailyCost > DAILY_COST_THRESHOLD) {
      const alertKey = `groq:alert:daily:${dateKey}`;
      const alreadyAlerted = await redis.get(alertKey);
      
      if (!alreadyAlerted) {
        console.warn(`[GROQ ALERT] Daily cost ($${dailyCost.toFixed(2)}) exceeds threshold ($${DAILY_COST_THRESHOLD})`);
        await redis.set(alertKey, "1", { EX: 24 * 60 * 60 }); // Alert once per day
      }
    }
    
    // Check monthly threshold
    if (monthlyCost > MONTHLY_COST_THRESHOLD) {
      const alertKey = `groq:alert:monthly:${monthKey}`;
      const alreadyAlerted = await redis.get(alertKey);
      
      if (!alreadyAlerted) {
        console.warn(`[GROQ ALERT] Monthly cost ($${monthlyCost.toFixed(2)}) exceeds threshold ($${MONTHLY_COST_THRESHOLD})`);
        await redis.set(alertKey, "1", { EX: 30 * 24 * 60 * 60 }); // Alert once per month
      }
    }
  } catch (error) {
    console.error("Failed to check alerts:", error);
  }
}

/**
 * Get usage statistics for the last N days
 */
export async function getUsageStats(days: number = 30): Promise<{
  periodDays: number;
  totalRequests: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalTokens: number;
  totalCost: number;
  dailyBreakdown: Array<{
    date: string;
    requests: number;
    cost: number;
  }>;
}> {
  try {
    const redis = getRedisClient();
    const now = new Date();
    const dailyBreakdown: Array<{ date: string; requests: number; cost: number }> = [];
    let totalRequests = 0;
    let totalInputTokens = 0;
    let totalOutputTokens = 0;
    let totalCost = 0;
    
    // Get usage for each day in the period
    for (let i = 0; i < days; i++) {
      const date = new Date(now);
      date.setDate(date.getDate() - i);
      const dateKey = date.toISOString().split('T')[0];
      
      const daily = await getDailyUsage(dateKey);
      dailyBreakdown.push({
        date: dateKey,
        requests: daily.requests,
        cost: daily.cost
      });
      
      totalRequests += daily.requests;
      totalInputTokens += daily.inputTokens;
      totalOutputTokens += daily.outputTokens;
      totalCost += daily.cost;
    }
    
    return {
      periodDays: days,
      totalRequests,
      totalInputTokens,
      totalOutputTokens,
      totalTokens: totalInputTokens + totalOutputTokens,
      totalCost: Math.round(totalCost * 1000000) / 1000000,
      dailyBreakdown: dailyBreakdown.reverse() // Oldest to newest
    };
  } catch (error) {
    console.error("Failed to get usage stats:", error);
    return {
      periodDays: days,
      totalRequests: 0,
      totalInputTokens: 0,
      totalOutputTokens: 0,
      totalTokens: 0,
      totalCost: 0,
      dailyBreakdown: []
    };
  }
}

